### ISSUE TITLE
Feature request: Emit py3kwarnings for division of integer arrays.

### ISSUE THREAD

**Issue by stuarteberg** [numpy/numpy#7949]
**Update:** @njase has now submitted PR #8083 to implement this feature.
---

To help people migrate their code bases from Python 2 to Python 3, the python interpreter has a handy option `-3`, which issues warnings at runtime.  One of its warnings is for integer division:

``` bash
$ echo "print 3/2" > /tmp/foo.py
$ python -3 /tmp/foo.py
/tmp/foo.py:1: DeprecationWarning: classic int division
  print 3/2
1
```

In Python 3, the new integer division semantics [also apply to numpy arrays](http://docs.scipy.org/doc/numpy/reference/generated/numpy.divide.html).  But currently, numpy does not emit warnings when such divisions are encountered, e.g. for a statement like this:

``` python
print np.array([3]) / np.array([2])
```

It would be very useful for numpy to provide py3kwarnings in such cases.

PS -- This issue was originally discussed in this mailing list thread:
https://mail.scipy.org/pipermail/numpy-discussion/2016-August/075897.html

**Comment by seberg** [comment]
Maybe you are even inclined to try to add it? It should only be moderately easy, you need to create a new type resolver in /core/src/umath which also adds a warning (checking the Py_Py3kWarnFlag on python 3) and then use that one in numpy/core/code_generators/generate_umath for the divide ufunc.
There are probably quite a few other places, would make sense to grep through checking for compile time switches between Python 2 and python 3 to figure out where it would make sense.

**Comment by njase** [comment]
HI, @seberg 
I was trying to understand the code to be able to fix this issue but I am stuck. Could you advise?

Here's the call flow I see during numpy division operation (core/src/multiarray/number.c):
array_true_divide() --> PyArray_GenericBinaryFunction() --> PyObject_CallFunctionObjArgs()

The first argument to PyObject_CallFunctionObjArgs(op, m1, m2, NULL) must be a callable python object. in our case, it is n_ops.true_divide

Now I don't understand (and can't find) what kind of object is "true_divide" and where is it defined in umath?

(Its my first attempt at numpy code..appreciate your help)

**Comment by njase** [comment]
Hi, @seberg 
Following your original comment about changing divide ufunc, I am able to get the desired behaviour. 
Few things:
1. This issue requests a behaviour of numpy which is similar to "python  -3", invoked under python 2.x environment, so the flag Py_DivisionWarningFlag should be used instead of Py_Py3kWarningFlag. This is also how python 2.x code does it
2. This flag is not used anywhere else in numpy code
3. After a grep, I find that other impacted places would be - "numpy/core/src/multiarray/number.c, 
numpy/core/tests/test_multiarray.py, numpy/ma/core.py, numpy/polynomial/_polybase.py". Fortunately, all of them use ndarry division operation at the end. So a generic change in divide ufunc should suffice

Let me know your view

**Comment by seberg** [comment]
Sorry, rather busy the last week. Sounds like you spot on.

If you add the warning and then run the test suit in master (with the flag), it might be rather annoying to see lots of tests failing. Please have a look at what are failing in any case. We could find bugs in numpy itself even, or may want to make sure there are no warnings when it does not matter or is deeply hidden.

**Comment by njase** [comment]
ok i finally used the modified numpy for my project (it did find issues :)
**On master test suite**: (Note: I am testing under python 2.7.12)
I tried python -3 with the usual test suite of numpy (logs attached)
- 4 cases of deprecated integer division found : These are in test case where several mathematical operations are tested all at once. I see this warning as harmless in such cases
- All other errors are of DepreciationWarning of different categories - To my understanding they exist because I am using python 2.7.x environment. They must've been resolved in numpy under python 3.x. So I am not inclined to work on these errors

**On new test cases**:
I think that I should add some test cases before submitting these changes to numpy repository but existing numpy cases are all invoked with standard python command line, without any additional arguments. 
It is possible to add a new case in which 

-  a subprocess of python is invoked with -3
-  only deprecated integer division is tested

[ilastik_op.txt](https://github.com/numpy/numpy/files/485475/ilastik_op.txt)

But somehow it does not look clean...there must be a better way

**Comment by seberg** [comment]
I am having a bit trouble finding the test suit issues there (maybe you attached the wrong file?), but could also run it myself some time. We have to fix/hide all of the issues for the test to run successfully, but I think we may be OK to hide some of them on the global level if necessary.

If it is just a few, we can suppress them more locally.

We probably need to add some flags to the `.travis` file and `numpy/tools/travis-test.sh` to run python in one instance with `-3`. If you do that, just create a pull request, then I can check the errors without even pulling myself.

**Comment by njase** [comment]
Oh, sorry for wrong file. I can only provide it and travis stuff tomorrow.
Thanks

**Comment by njase** [comment]
you can see the results here: https://travis-ci.org/numpy/numpy/builds/162168742

**Comment by stuarteberg** [comment]
This feature was implemented in #8083, which was merged today.  Hooray!
